src/stk/databases/mongo_db/constructed_molecule.py

```diff
@@ -331,32 +352,130 @@ def make_hashable(json):
-        # in the lru_cache is modified with some extra items added by
-        # insert_one(). This means that the next time _put() is used
-        # with a clean json, it will not match the one in the cache,
-        # because the one in the cache has the extra items added by
-        # insert_one(). To prevent this use
-        # insert_one(dict(json['molecule'])), which means that a copy
-        # is modified by insert_one and the json in the cache is
-        # not changed.
-
-        self._position_matrices.insert_one(dict(json['matrix']))
-        self._molecules.insert_one(dict(json['molecule']))
-        self._constructed_molecules.insert_one(
-            document=dict(json['constructedMolecule']),
+        query = self._get_query(json)
+        self._molecules.update_many(
+            filter=query,
+            update={
+                '$set': json['molecule'],
+            },
+            upsert=True,
+        )
+        self._position_matrices.update_many(
+            filter=query,
+            update={
+                '$set': json['matrix'],
+            },
+            upsert=True,
+        )
+
+        self._add_building_block_keys_from_database(
+            query=query,
+            building_block_keys=json['constructedMolecule']['BB'],
+        )
+
+        self._constructed_molecules.update_many(
+            filter=query,
+            update={
+                '$set': json['constructedMolecule'],
+            },
+            upsert=True,
         )
         for building_block_json in json['buildingBlocks']:
-            self._molecules.insert_one(
-                document=dict(building_block_json['molecule']),
+            building_block_query = self._get_query(building_block_json)
+            self._molecules.update_many(
+                filter=building_block_query,
+                update={
+                    '$set': building_block_json['molecule'],
+                },
+                upsert=True,
             )
-            self._position_matrices.insert_one(
-                document=dict(building_block_json['matrix']),
+            self._building_block_position_matrices.update_many(
+                filter=building_block_query,
+                update={
+                    '$set': building_block_json['matrix'],
+                },
+                upsert=True,
             )
 
+    def _add_building_block_keys_from_database(
+        self,
+        query,
+        building_block_keys,
+    ):
+        """
+        Add previously deposited keys to `building_block_keys`.
+
+        Checks the constructed molecule collection to find all
+        constructed molecule entries which match `query`. All matches
+        should merely be duplicate entries for the same constructed
+        molecule.
+
+        Each entry for the constructed molecule will have a
+        :class:`list` of building blocks, which were used to
+        construct the constructed molecule.
```

Comment:
I am assuming their defined positions in the constructed molecule are also held somewhere (i.e. the topology graph)? Otherwise, this is required because you cannot reliably recreate all constructions without knowing all of this information.

Reply:
The topology graph is not a part of the ConstructedMolecule -> it is not deposited in to the database as is not needed to restore a ConstructedMolecule.